### DISCUSSION TITLE
Retrieving module output via RPC client

### DISCUSSION THREAD

**Discussion by EmilioPanti** [rapid7/metasploit-framework#15517] [Q&A] [answered]
I am using metasploit via RPC server and RPC client.
Is there a way to retrieve the output of a post module executed via the `module.execute` call? 
By output I mean the text that would be displayed on the terminal when that module is run via msfconsole

**Comment by XZGLC** [comment]
You had better connect to a VPN which is accessible to Google. And to remember that set the time zone and language of your computer same to which in the US.

**Reply by EmilioPanti** [reply]
Maybe I explained myself wrong.
I have no problem communicating with the RPC servers from the RPC client.
My question was if there was a way to retrieve the output generated by running a module via some client RPC specific calls

**Comment by timwr** [comment, marked as answer]
Unfortunately I think you have open a new console and manually run the module with `console.write` (e.g use module; set option1 blah; run). I think is what armitage used to do. Potentially this could be improved by modifying the rpc api.

**Reply by EmilioPanti** [reply]
I was afraid that the answer was this.

Assuming that I run modules via `console.write` could I continue to use the calls `session.shell_write`, `session.meterpreter_write`, etc to communicate with open sessions, or would I be forced to use `console.write` to communicate with sessions as well?
I would not want sessions opened via console commands (`console.write`) to be invisible to other api calls that do not go "through" the console.

Session management with `console.write` would be rather annoying ... it would be very useful to have an RPC call to retrieve outputs/logs of modules executed with the api `module.execute`

**Reply by timwr** [reply]
Yeah that's fine I think. You may need to run the module as a job `run -j` to ensure you don't automatically interact with the new session?

**Reply by EmilioPanti** [reply]
I suppose so, I'll do some experiments. Thanks for the replies!

**Reply by EmilioPanti** [reply]
@timwr I tried to investigate a bit in the framework code to see if I could somehow make changes in order to save the output of a job (which are created when you run `module.execute`) in some text file on the server or in memory but unfortunately I am not familiar with either ruby or the general structure of the framework.

Do you have any suggestions or hints to accomplish this?
